exec/internal/compile/backend_amd64.go

```diff
@@ -0,0 +1,311 @@
+// Copyright 2019 The go-interpreter Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package compile
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	ops "github.com/go-interpreter/wagon/wasm/operators"
+	asm "github.com/twitchyliquid64/golang-asm"
+	"github.com/twitchyliquid64/golang-asm/obj"
+	"github.com/twitchyliquid64/golang-asm/obj/x86"
+)
+
+// dirtyRegs hold booleans that are true when the register stores
+// a reserved value that needs to be flushed to memory.
+type dirtyRegs struct {
+	R12 bool
+	R13 bool
+}
+
+// Details of the AMD64 backend:
+// Reserved registers (for now):
+//  - R10 - pointer to stack sliceHeader
+//  - R11 - pointer to locals sliceHeader
+//  - R12 - pointer for stack item
+//  - R13 - stack size
+// Scratch registers:
+//  - RAX, RBX, RCX, RDX, R8, R9, R15
+// Most emission instructions make few attempts to optimize in order
+// to keep things simple, however a planned second pass peephole-optimizer
+//  should make a big difference.
+
+// AMD64Backend is the native compiler backend for x86-64 architectures.
+type AMD64Backend struct {
+	s *scanner
+}
+
+// Scanner returns a scanner that can be used for
+// emitting compilation candidates.
+func (b *AMD64Backend) Scanner() *scanner {
+	if b.s == nil {
+		b.s = &scanner{
+			supportedOpcodes: map[byte]bool{
+				ops.I64Const: true,
+				ops.I64Add:   true,
+				ops.I64Sub:   true,
+				ops.I64And:   true,
+				ops.I64Or:    true,
+				ops.I64Mul:   true,
+				ops.GetLocal: true,
+			},
+		}
+	}
+	return b.s
+}
+
+// Build implements exec.instructionBuilder.
+func (b *AMD64Backend) Build(candidate CompilationCandidate, code []byte, meta *BytecodeMetadata) ([]byte, error) {
+	builder, err := asm.NewBuilder("amd64", 128)
+	if err != nil {
+		return nil, err
+	}
+	var regs dirtyRegs
+	b.emitPreamble(builder, &regs)
+
+	for i := candidate.StartInstruction; i <= candidate.EndInstruction; i++ {
+		//fmt.Printf("i=%d, meta=%+v, len=%d\n", i, meta.Instructions[i], len(code))
+		inst := meta.Instructions[i]
+		switch inst.Op {
+		case ops.I64Const:
+			b.emitPushI64(builder, &regs, b.readIntImmediate(code, inst))
+		case ops.GetLocal:
+			b.emitWasmLocalsLoad(builder, &regs, x86.REG_AX, b.readIntImmediate(code, inst))
+			b.emitWasmStackPush(builder, &regs, x86.REG_AX)
+		case ops.I64Add, ops.I64Sub, ops.I64Mul, ops.I64Or, ops.I64And:
+			if err := b.emitBinaryI64(builder, &regs, inst.Op); err != nil {
+				return nil, fmt.Errorf("emitBinaryI64: %v", err)
+			}
+		default:
+			return nil, fmt.Errorf("cannot handle inst[%d].Op 0x%x", i, inst.Op)
+		}
+	}
+	b.emitPostamble(builder, &regs)
+
+	out := builder.Assemble()
+	// cmd := exec.Command("ndisasm", "-b64", "-")
```

Comment:
there are many of these snippets of code being commented out.
I understand they were quite useful to debug things.

perhaps we should replace them with a proper helper func `ndisasm`, guarded with either a debug flag or an environment variable?

or just remove them :)

Reply:
I'm a fan of leaving them in as they are *very* useful :), and to a new contributor it may be helpful for them to see what is happening. I'll change them all to calls to `debugPrintAsm()`.